x-pack/plugins/alerting/server/alerts_client/lib/build_recovered_alert.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEmpty } from 'lodash';
+import type { Alert } from '@kbn/alerts-as-data-utils';
+import { Alert as LegacyAlert } from '../../alert/alert';
+import { AlertInstanceContext, AlertInstanceState, RuleAlertData } from '../../types';
+import type { AlertRule } from '../types';
+
+interface BuildRecoveredAlertOpts<
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  alert: Alert & AlertData;
+  legacyAlert: LegacyAlert<LegacyState, LegacyContext, ActionGroupIds | RecoveryActionGroupId>;
+  rule: AlertRule;
+  timestamp: string;
+}
+
+/**
+ * Updates an existing alert document with data from the LegacyAlert class
+ * This could be a currently active alert that is now recovered or a previously
+ * recovered alert that has updates to its flapping history
+ * Currently only populates framework fields and not any rule type specific fields
+ */
+
+export const buildRecoveredAlert = <
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>({
+  alert,
+  legacyAlert,
+  rule,
+  timestamp,
+}: BuildRecoveredAlertOpts<
+  AlertData,
+  LegacyState,
+  LegacyContext,
+  ActionGroupIds,
+  RecoveryActionGroupId
+>): Alert & AlertData => {
+  return {
+    ...alert,
+    // Update the timestamp to reflect latest update time
+    '@timestamp': timestamp,
+    kibana: {
+      ...alert.kibana,
+      alert: {
+        ...alert.kibana.alert,
```

Comment:
Good catch! Updated in this commit: https://github.com/elastic/kibana/pull/156946/commits/52e1b40eaf963f9cf3363f461ae4d388382ca563

I also updated the logic for new and ongoing alerts since I realized that the `maintenance_window_ids` array can be empty during an execution and that empty array should overwrite any existing maintenance window IDs that may have been persisted in previous executions.